state/protocol/protocol_state/mutator_test.go

```diff
@@ -1,72 +1,289 @@
+	require.Empty(s.T(), dbUpdates)
 }
 
-// TestMutatorHappyPathNoChanges tests that Mutator correctly indexes the protocol state when there are no changes.
-func (s *MutatorSuite) TestMutatorHappyPathNoChanges() {
+// TestHappyPathHasChanges tests that stateMutator returns cached db updates when building protocol state after applying service events.
+func (s *StateMutatorSuite) TestHappyPathHasChanges() {
+	s.params.On("EpochFallbackTriggered").Return(false, nil)
 	parentState := unittest.ProtocolStateFixture()
-	candidate := unittest.BlockHeaderFixture(unittest.HeaderWithView(parentState.CurrentEpochSetup.FirstView))
-	s.protocolStateDB.On("ByBlockID", candidate.ParentID).Return(parentState, nil)
-	updater, err := s.mutator.CreateUpdater(candidate)
-	require.NoError(s.T(), err)
+	s.stateMachine.On("ParentState").Return(parentState)
+	s.stateMachine.On("Build").Return(unittest.ProtocolStateFixture().ProtocolStateEntry,
+		unittest.IdentifierFixture(), true)
 
-	s.protocolStateDB.On("Index", candidate.ID(), parentState.ID()).Return(func(tx *transaction.Tx) error { return nil })
+	epochSetup := unittest.EpochSetupFixture()
+	result := unittest.ExecutionResultFixture(func(result *flow.ExecutionResult) {
+		result.ServiceEvents = []flow.ServiceEvent{epochSetup.ServiceEvent()}
+	})
 
-	err = s.mutator.CommitProtocolState(updater)(&transaction.Tx{})
+	block := unittest.BlockHeaderFixture()
+	seal := unittest.Seal.Fixture(unittest.Seal.WithBlockID(block.ID()))
+	s.headersDB.On("ByBlockID", seal.BlockID).Return(block, nil)
+	s.resultsDB.On("ByID", seal.ResultID).Return(result, nil)
+
+	s.stateMachine.On("ProcessEpochSetup", epochSetup).Return(nil).Once()
+	s.setupsDB.On("StoreTx", epochSetup).Return(func(*transaction.Tx) error { return nil }).Once()
+
+	err := s.mutator.ApplyServiceEvents([]*flow.Seal{seal})
 	require.NoError(s.T(), err)
+
+	_, _, _, dbUpdates := s.mutator.Build()
+	require.NoError(s.T(), err)
+	require.Len(s.T(), dbUpdates, 1)
 }
 
-// TestMutatorHappyPathHasChanges tests that Mutator correctly persists and indexes the protocol state when there are changes.
-func (s *MutatorSuite) TestMutatorHappyPathHasChanges() {
+// TestApplyServiceEvents_InvalidEpochSetup tests that handleServiceEvents rejects invalid epoch setup event and sets
+// InvalidStateTransitionAttempted flag in protocol.ProtocolStateMachine.
+func (s *StateMutatorSuite) TestApplyServiceEvents_InvalidEpochSetup() {
+	s.params.On("EpochFallbackTriggered").Return(false, nil)
+	s.Run("invalid-epoch-setup", func() {
+		parentState := unittest.ProtocolStateFixture()
+		rootSetup := parentState.CurrentEpochSetup
+
+		s.stateMachine.On("ParentState").Return(parentState)
+
+		epochSetup := unittest.EpochSetupFixture(
+			unittest.WithParticipants(rootSetup.Participants),
+			unittest.SetupWithCounter(rootSetup.Counter+1),
+			unittest.WithFinalView(rootSetup.FinalView+1000),
+			unittest.WithFirstView(rootSetup.FinalView+1),
+		)
```

Comment:
We might not need all this, since we're mocking the validation.
```suggestion
		epochSetup := unittest.EpochSetupFixture()
```

Reply:
Yesno, we can omit it once we move ALL validation in `ProcessEpochSetup`, in fact I have already done this but on separate branch. For now it has to stay like this since we do validation in `ApplyServiceEvents`